### PR TITLE
add cluster autoscaler addon

### PR DESCRIPTION
**Pull request by hakuna-matatah**
*Description of changes:*
add cluster autoscaler addon.
Clusteraddon tests can be enabled like any other addon by setting env variable
It's disabled by default.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: eks/ng/role.go [hunk 1]
```diff
@@ -205,8 +208,30 @@ Outputs:
...
+            - ec2:DescribeLaunchTemplateVersions
+            Resource: "*"
+`
+
 func (ts *tester) createRole() error {
 	if !ts.cfg.EKSConfig.AddOnNodeGroups.RoleCreate {
+		// if role isn't created and autoscaler is enabled then return error
+		// because we cannot assume ASG policy is attached to node iam role
+		// this check is also enabled when calling createCA() in nodes.go which comes after createRole() func call
+		if ts.cfg.EKSConfig.IsEnabledAddOnClusterAutoscaler() {
```
Comment (gyuho) [inline eks/ng/role.go:232]: Remove `ts.cfg.EKSConfig.IsEnabledAddOnClusterAutoscaler` since we are adding autoscaler inside node group `struct`. You can iterate `NGs` and store a boolean variable to indicate whether to add ASG policy or not.

File: eksconfig/add-on-cluster-autoscaler.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package eksconfig
```
Comment (gyuho) [inline eksconfig/add-on-cluster-autoscaler.go:1]: This should be inside `AddOnNodeGroups`

File: eksconfig/config.go [hunk 3]
```diff
@@ -303,9 +303,12 @@ type Config struct {
 	// AddOnClusterVersionUpgrade defines parameters
 	// for EKS cluster version upgrade add-on.
 	AddOnClusterVersionUpgrade *AddOnClusterVersionUpgrade `json:"add-on-cluster-version-upgrade,omitempty"`
+	// AddOnClusterAutoscaler defines parameters for EKS Cluster
+	// add on clusterautoscaler
+	AddOnClusterAutoscaler *AddOnClusterAutoscaler `json:"add_on_cluster_autoscaler,omitempty"`
```
[the remark this answers is not in the document]
Comment (gyuho) [reply eksconfig/config.go:308]: That way, we can configure cluster auto scaler "per" node group. See how we implement `version-upgrade` for MNG

**Comment by gyuho** [comment]
You don't need create a new PR. You can push the changes to the same branch.

File: eksconfig/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gyuho) [reply eksconfig/config.go:308]: Similar to https://github.com/aws/aws-k8s-tester/issues/91.

File: eks/ng/role.go [hunk 4]
```diff
@@ -196,6 +198,7 @@ Resources:
             - route53:ChangeResourceRecordSets
             - route53:DeleteHealthCheck
             Resource: "*"
+{{ if ne .UserData "" }}{{.UserData}}{{ end }}
```
[the remark this answers is not in the document]
Comment (hakuna-matatah) [reply eks/ng/role.go:201]: I can rename but UserData is not being used in this file at all ( here there is no EC2 bootstrap data for role)

File: eksconfig/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hakuna-matatah) [reply eksconfig/config.go:308]: Ah! per node group. More flexibility than enabling CA for every node group. I like it. I will do that.

File: eks/ng/role.go (hunk 1 above)
Comment (hakuna-matatah) [reply eks/ng/role.go:232]: Sure, that makes more sense ( as we are planning to move it inside the NG)

File: eksconfig/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hakuna-matatah) [reply eksconfig/config.go:308]: > ```
> type AddOnNodeGroups struct {
>     ...
>     AutoScaler *NGAutoScaler `json:"auto-scaler"`
>     ....
> ```
> 
> ```
> type NGAutoScaler struct {
>     Enable bool `json:"enable"`
>     ...
> ```

Actually it makes sense to add to ASG Struct than AddonNodeGroupStruct 
```
type NGClusterAutoScaler struct {
	Enable bool `json:"enable"`
	// Created is true when the resource has been created.
	// Used for delete operations.
	Created         bool               `json:"created" read-only:"true"`
	TimeFrameCreate timeutil.TimeFrame `json:"time-frame-create" read-only:"true"`
}
```
```
type ASG struct {
	...
	ClusterAutoScaler *NGClusterAutoScaler `json:"cluster_auto_scaler,omitempty"`
...
}
```

File: eksconfig/add-on-cluster-autoscaler.go (hunk 2 above)
Comment (hakuna-matatah) [reply eksconfig/add-on-cluster-autoscaler.go:1]: > This should be inside `AddOnNodeGroups`

This doesn't make sense anymore as a separate file if we are planning to take the route of embedding inside NodeGroup/ASG.

File: eks/ng/cluster-autoscaler/cluster-autoscaler.go [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+
+type tester struct {
+	cfg Config
+}
+
+func (ts *tester) CreateCA() error {
+	return ts.createCA()
+}
+
+func (ts *tester) createCA() error {
```
Comment (gyuho) [inline eks/ng/cluster-autoscaler/cluster-autoscaler.go:250]: Can we iterate all `ts.cfg.EKSConfig.AddOnNodeGroups.ASGs` and check if we need install CA?

```
needInstall := false
for ... range ts.cfg.EKSConfig.AddOnNodeGroups.ASGs {
   if cur.Autoscaler.Enable { needInstall = true; break; }
   ...

if !needInstall { lg.Info("no NG enables CA; skipping"); return nil; }
```

File: eksconfig/README.md [hunk 6]
```diff
@@ -36,6 +36,7 @@ AWS_K8S_TESTER_EKS_ADD_ON_STRESSER_REMOTE_ENABLE=true \
 AWS_K8S_TESTER_EKS_ADD_ON_CLUSTER_LOADER_LOCAL_ENABLE=true \
 AWS_K8S_TESTER_EKS_ADD_ON_CLUSTER_LOADER_REMOTE_ENABLE=true \
 AWS_K8S_TESTER_EKS_ADD_ON_CLUSTER_VERSION_UPGRADE_ENABLE=true \
+AWS_K8S_TESTER_ADD_ON_CLUSTER_AUTOSCALER_ENABLE=true \
```
Comment (gyuho) [inline eksconfig/README.md:39]: We do not want to cluster-wide autoscaler enable. This should be scoped down to individual node groups. I think your code already does that. Can we remove this?

File: eksconfig/add-on-node-groups.go [hunk 7]
```diff
@@ -68,6 +76,8 @@ type ASG struct {
 	// TODO: handle conflicting flag '--cloud-provider aws'
 	// ref. https://github.com/kubernetes/kubernetes/issues/64659
 	KubeletExtraArgs string `json:"kubelet-extra-args"`
+	// clusterautoscaler per nodegroup
+	ClusterAutoScaler *NGClusterAutoScaler `json:"cluster_auto_scaler,omitempty"`
```
Comment (gyuho) [inline eksconfig/add-on-node-groups.go:80]: To be consistent with other JSON tags, let's just name it `json:"cluster-autoscaler,omitempty"`

File: eksconfig/env_test.go [hunk 8]
```diff
@@ -141,7 +141,7 @@ func TestEnv(t *testing.T) {
...
+	os.Setenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ASGS", `{"ng-test-name-cpu":{"name":"ng-test-name-cpu","remote-access-user-name":"ec2-user","ami-type":"AL2_x86_64","image-id-ssm-parameter":"/aws/service/eks/optimized-ami/1.30/amazon-linux-2/recommended/image_id","asg-min-size":17,"kubelet-extra-args":"bbb qq", "cluster_auto_scaler" : {"enable" : false}, asg-max-size":99,"asg-desired-capacity":77,"instance-types":["type-cpu-2"],"volume-size":40},"ng-test-name-gpu":{"name":"ng-test-name-gpu","remote-access-user-name":"ec2-user","ami-type":"AL2_x86_64_GPU","asg-min-size":30,"asg-max-size":35,"asg-desired-capacity":34,"instance-types":["type-gpu-2"],"image-id":"my-gpu-ami","volume-size":500, "cluster_auto_scaler": {"enable":false},"kubelet-extra-args":"aaa aa"}}`)
```
Comment (gyuho) [inline eksconfig/env_test.go:144]: This should be `cluster-autoscaler`.

File: eksconfig/env_test.go [hunk 9]
```diff
@@ -1352,7 +1354,7 @@ func TestEnvAddOnNodeGroupsGetRef(t *testing.T) {
 
 	os.Setenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ENABLE", `true`)
 	defer os.Unsetenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ENABLE")
-	os.Setenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ASGS", `{"GetRef.Name-ng-for-cni":{"name":"GetRef.Name-ng-for-cni","remote-access-user-name":"ec2-user","ami-type":"AL2_x86_64","asg-min-size":30,"asg-max-size":35,"asg-desired-capacity":34,"image-id":"my-ami",  "ssm-document-create":true,   "instance-types":["type-2"],  "ssm-document-cfn-stack-name":"GetRef.Name-ssm", "ssm-document-name":"GetRef.Name-document",     "kubelet-extra-args":"aaa aa",  "volume-size":500}}`)
+	os.Setenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ASGS", `{"GetRef.Name-ng-for-cni":{"name":"GetRef.Name-ng-for-cni","remote-access-user-name":"ec2-user","ami-type":"AL2_x86_64","asg-min-size":30,"asg-max-size":35,"asg-desired-capacity":34,"image-id":"my-ami",  "ssm-document-create":true,   "instance-types":["type-2"],  "ssm-document-cfn-stack-name":"GetRef.Name-ssm", "ssm-document-name":"GetRef.Name-document", "kubelet-extra-args":"aaa aa", "cluster_auto_scaler": {"enable" : false}, "volume-size":500}}`)
```
Comment (gyuho) [inline eksconfig/env_test.go:1357]: `cluster-autoscaler`

File: eksconfig/env_test.go [hunk 10]
```diff
@@ -1426,7 +1429,7 @@ func TestEnvAddOnConformance(t *testing.T) {
 
 	os.Setenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ENABLE", `true`)
 	defer os.Unsetenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ENABLE")
-	os.Setenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ASGS", `{"GetRef.Name-ng-for-cni":{"name":"GetRef.Name-ng-for-cni","remote-access-user-name":"ec2-user","ami-type":"AL2_x86_64","asg-min-size":30,"asg-max-size":35,"asg-desired-capacity":34,"image-id":"my-ami",  "ssm-document-create":true,   "instance-types":["type-2"],  "ssm-document-cfn-stack-name":"GetRef.Name-ssm", "ssm-document-name":"GetRef.Name-document",     "kubelet-extra-args":"aaa aa",  "volume-size":500}}`)
+	os.Setenv("AWS_K8S_TESTER_EKS_ADD_ON_NODE_GROUPS_ASGS", `{"GetRef.Name-ng-for-cni":{"name":"GetRef.Name-ng-for-cni","remote-access-user-name":"ec2-user","ami-type":"AL2_x86_64","asg-min-size":30,"asg-max-size":35,"asg-desired-capacity":34,"image-id":"my-ami",  "ssm-document-create":true,   "instance-types":["type-2"],  "ssm-document-cfn-stack-name":"GetRef.Name-ssm", "ssm-document-name":"GetRef.Name-document", "cluster_auto_scaler": {"enable" : false}, "kubelet-extra-args":"aaa aa",  "volume-size":500}}`)
```
Comment (gyuho) [inline eksconfig/env_test.go:1432]: `cluster-autoscaler`

File: eksconfig/add-on-node-groups.go [hunk 11]
```diff
@@ -56,6 +56,14 @@ type AddOnNodeGroups struct {
 	ASGs map[string]ASG `json:"asgs,omitempty"`
 }
 
+type NGClusterAutoScaler struct {
+	Enable bool `json:"enable"`
+	// Created is true when the resource has been created.
+	// Used for delete operations.
+	Created         bool               `json:"created" read-only:"true"`
+	TimeFrameCreate timeutil.TimeFrame `json:"time-frame-create" read-only:"true"`
```
Comment (gyuho) [inline eksconfig/add-on-node-groups.go:64]: Can we add another field `Image string`, and set default value.

File: eksconfig/README.md (hunk 6 above)
Comment (hakuna-matatah) [reply eksconfig/README.md:39]: > We do not want to cluster-wide autoscaler enable. This should be scoped down to individual node groups. I think your code already does that. Can we remove this?

forgot t remove it.

File: eksconfig/add-on-node-groups.go (hunk 11 above)
Comment (hakuna-matatah) [reply eksconfig/add-on-node-groups.go:64]: We are not doing this for now. Giving that option would lead into other problems as supplying wrong image that's not in GCR etc.

File: eks/ng/autoscaler/cluster-autoscaler.go [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+	}
+	return ts.createCA()
+}
+
+func (ts *tester) createCA() error {
+	ts.cfg.Logger.Info("creating CA using kubectl", zap.String("name", ts.cfg.EKSConfig.Name))
+	for asgName, cur := range ts.cfg.EKSConfig.AddOnNodeGroups.ASGs {
+		createStart := time.Now()
+		ts.cfg.Logger.Info("creating CA", zap.String("name", asgName))
+		if err := ts.installCA(); err != nil {
```
Comment (gyuho) [inline eks/ng/autoscaler/cluster-autoscaler.go:268]: `InstallCA` should take `cur.Name` as a parameter, otherwise we install the same plugin x number of times?

File: eks/ng/autoscaler/cluster-autoscaler.go [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+			return err
+		}
+		ts.cfg.Logger.Info("created CA",
+			zap.String("asg-name", cur.Name),
+			zap.String("started", humanize.RelTime(createStart, time.Now(), "ago", "from now")),
+		)
+		ts.cfg.EKSConfig.AddOnNodeGroups.ASGs[asgName] = cur
+		_ = ts.cfg.EKSConfig.Sync()
```
Comment (gyuho) [inline eks/ng/autoscaler/cluster-autoscaler.go:276]: We can just ignore without `_=`?

File: eks/ng/autoscaler/cluster-autoscaler.go [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+		createStart := time.Now()
+		ts.cfg.Logger.Info("creating CA", zap.String("name", asgName))
+		if err := ts.installCA(); err != nil {
+			return err
+		}
+		ts.cfg.Logger.Info("created CA",
+			zap.String("asg-name", cur.Name),
+			zap.String("started", humanize.RelTime(createStart, time.Now(), "ago", "from now")),
+		)
+		ts.cfg.EKSConfig.AddOnNodeGroups.ASGs[asgName] = cur
```
Comment (gyuho) [inline eks/ng/autoscaler/cluster-autoscaler.go:275]: Actually, remove line 275 and 276. Doesn't seem like we change anything in `cur`?

File: eks/ng/autoscaler/cluster-autoscaler.go [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+		_ = ts.cfg.EKSConfig.Sync()
+	}
+	return ts.cfg.EKSConfig.Sync()
+}
+
+func (ts *tester) installCA() error {
+	var caData = caSpecData{}
+	caData.ImageURI = images[ts.cfg.EKSConfig.Parameters.Version]
```
Comment (gyuho) [inline eks/ng/autoscaler/cluster-autoscaler.go:283]:
```
caData.ImageURI, ok = images[...]
if !ok { return fmt.Errorf("no CA found for %q", ...Version) }
```

File: eks/ng/autoscaler/cluster-autoscaler.go [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+	}
+	if err != nil {
+		return fmt.Errorf("'kubectl apply' failed %v (output %q)", err, string(output))
+	}
+	ts.cfg.Logger.Info("created cluster autoscaler")
+
+	return ts.waitDeploymentCA()
+}
+
+func (ts *tester) waitDeploymentCA() error {
```
Comment (gyuho) [inline eks/ng/autoscaler/cluster-autoscaler.go:340]: This should take `clusterAutoScalerDeploymentName` as a parameter, in case we run multiple cluster autoscalers for multiple node groups?

File: eks/ng/role.go [hunk 17]
```diff
@@ -238,8 +257,14 @@ func (ts *tester) createRole() error {
 	if ts.cfg.EKSConfig.AddOnNodeGroups.RoleName == "" {
 		return errors.New("cannot create a cluster role with an empty AddOnNodeGroups.RoleName")
 	}
-
-	if err := ioutil.WriteFile(ts.cfg.EKSConfig.AddOnNodeGroups.RoleCFNStackYAMLFilePath, []byte(TemplateRole), 0400); err != nil {
+	tr := templateRole{}
+	tr.ASGPolicy = asgPolicyData
```
Comment (gyuho) [inline eks/ng/role.go:261]: Only set this when autoscaler is enabled?

File: eksconfig/add-on-node-groups.go [hunk 18]
```diff
@@ -232,6 +243,10 @@ func (cfg *Config) validateAddOnNodeGroups() error {
 			return fmt.Errorf("%q both ImageID and ImageIDSSMParameter are empty", cur.Name)
 		}
 
+		if cur.ClusterAutoScaler.Enable && !cfg.AddOnNodeGroups.RoleCreate {
```
Comment (gyuho) [inline eksconfig/add-on-node-groups.go:246]: Also error if `CA.Enable == true` and `AddOnMetricsServer.Enable == false`? Is metrics-server requirement for CA?

File: eks/ng/role.go [hunk 19]
```diff
@@ -196,6 +198,7 @@ Resources:
             - route53:ChangeResourceRecordSets
             - route53:DeleteHealthCheck
             Resource: "*"
+{{ if ne .ASGPolicy "" }}{{.ASGPolicy}}{{ end }}
```
Comment (gyuho) [inline eks/ng/role.go:201]: Can we add a simple test when `ASGPolicy` is not empty?

https://github.com/aws/aws-k8s-tester/blob/3e23f03490ef60feceec33f37e78b1996f39f6b9/eks/cluster/cluster_test.go#L10

File: eks/ng/autoscaler/cluster-autoscaler.go (hunk 12 above)
Comment (hakuna-matatah) [reply eks/ng/autoscaler/cluster-autoscaler.go:268]: > `InstallCA` should take `cur.Name` as a parameter, otherwise we install the same plugin x number of times?

I don't think it will install the same plugin. We are only installing once per cluster as long as one NG is enabled.

File: eks/ng/autoscaler/cluster-autoscaler.go (hunk 13 above)
Comment (hakuna-matatah) [reply eks/ng/autoscaler/cluster-autoscaler.go:276]: My IDE was showing as a warning :D ...so just added it. I think that should be fine ?

File: eksconfig/add-on-node-groups.go (hunk 18 above)
Comment (hakuna-matatah) [reply eksconfig/add-on-node-groups.go:246]: > Also error if `CA.Enable == true` and `AddOnMetricsServer.Enable == false`? Is metrics-server requirement for CA?

metrics-server is not a requirement for CA. Metrics server is a requirement for HPA.

File: eks/ng/autoscaler/cluster-autoscaler.go (hunk 16 above)
Comment (hakuna-matatah) [reply eks/ng/autoscaler/cluster-autoscaler.go:340]: cluster-autoscaler deployment will only be one per cluster.

File: eks/ng/role.go (hunk 17 above)
Comment (hakuna-matatah) [reply eks/ng/role.go:261]: I think this should be ok ? Because by default these policies will be included to the role if the node-groups (or) clusters are created using "eksctl".

File: eks/ng/role.go (hunk 19 above)
Comment (hakuna-matatah) [reply eks/ng/role.go:201]: > Can we add a simple test when `ASGPolicy` is not empty?
> 
> https://github.com/aws/aws-k8s-tester/blob/3e23f03490ef60feceec33f37e78b1996f39f6b9/eks/cluster/cluster_test.go#L10

Because we are adding ASGPolicy with value in any case, do we really care ?

File: eksconfig/add-on-node-groups.go [hunk 20]
```diff
@@ -68,6 +76,8 @@ type ASG struct {
 	// TODO: handle conflicting flag '--cloud-provider aws'
 	// ref. https://github.com/kubernetes/kubernetes/issues/64659
 	KubeletExtraArgs string `json:"kubelet-extra-args"`
+	// clusterautoscaler per nodegroup
+	ClusterAutoScaler *NGClusterAutoScaler `json:"cluster-autoscaler,omitempty"`
```
Comment (gyuho) [inline eksconfig/add-on-node-groups.go:80]: Let's move this inside `AddOnNodeGroups`

File: eks/ng/autoscaler/cluster-autoscaler.go (hunk 16 above)
Comment (gyuho) [reply eks/ng/autoscaler/cluster-autoscaler.go:340]: Ok makes sense.

File: eks/ng/role.go (hunk 17 above)
Comment (gyuho) [reply eks/ng/role.go:261]: ack

File: eksconfig/add-on-node-groups.go (hunk 18 above)
Comment (gyuho) [reply eksconfig/add-on-node-groups.go:246]: ack

File: eks/ng/autoscaler/cluster-autoscaler.go (hunk 13 above)
Comment (gyuho) [reply eks/ng/autoscaler/cluster-autoscaler.go:276]: ok let's keep it then

File: eks/ng/autoscaler/cluster-autoscaler.go (hunk 12 above)
Comment (gyuho) [reply eks/ng/autoscaler/cluster-autoscaler.go:268]: > for asgName, cur := range ts.cfg.EKSConfig.AddOnNodeGroups.ASGs {

still looping through this for-loop?  now that we are moving cluster autoscaler installer out, we can replace `createCA` with `installCA`?

Comment (gyuho) [reply eks/ng/autoscaler/cluster-autoscaler.go:268]: Talked offline. We still need per-ASG updates for tagging

File: eksconfig/add-on-node-groups.go (hunk 20 above)
Comment (hakuna-matatah) [reply eksconfig/add-on-node-groups.go:80]: talked offline. There is value in having it inside ASG. We can enable/disable CA per nodegroup.
